src/main/resources/assets/admin/common/js/ui/treegrid/TreeGrid.ts

```diff
@@ -2081,4 +1730,100 @@ export class TreeGrid<DATA>
         this.grid.selectRow(row, debounce);
     }
 
+    hasNodeWithDataId(dataId: string): boolean {
+        return !!this.root.getNodeByDataId(dataId);
+    }
+
+    protected insertDataToParentNode(data: DATA, parent: TreeNode<DATA>, index: number) {
+        const nodeToInsert: TreeNode<DATA> = this.dataToTreeNode(data, parent);
+        this.insertNodeToParentNode(nodeToInsert, parent, index);
+    }
+
+    protected appendDataToParentNode(data: DATA, parent: TreeNode<DATA>) {
+        const insertIndex: number = parent.getChildren().length;
+        this.insertDataToParentNode(data, parent, insertIndex);
+    }
+
+    protected insertNodeToParentNode(nodeToInsert: TreeNode<DATA>, parent: TreeNode<DATA>, index: number) {
+        parent.insertChild(nodeToInsert, index);
+        parent.setExpandable(true);
+
+        const parentRow: number = this.gridData.getRowById(parent.getId());
+        if (!!parentRow || parentRow === 0) {
+            this.gridData.insertItem(this.getInsertIndexRelativeToParent(parent, index), nodeToInsert);
+        }
+
+        this.invalidateNodes([parent]);
+    }
+
+    private getInsertIndexRelativeToParent(parent: TreeNode<DATA>, index: number): number {
+        const parentRow: number = this.gridData.getRowById(parent.getId());
+        let insertIndexRelativeToGrid: number = parentRow + 1;
+
+        let i: number = 0;
+        while (i !== index) {
+            if (this.gridData.getItem(insertIndexRelativeToGrid).getParent() === parent) {
+                i++;
+            }
+            insertIndexRelativeToGrid++;
+        }
+
+        return insertIndexRelativeToGrid;
+    }
+
+    moveNode(from: number, to: number): number {
+        const root: TreeNode<DATA> = this.getRoot().getCurrentRoot();
+        const rootChildren: TreeNode<DATA>[] = root.treeToList();
+
+        const item: TreeNode<DATA> = rootChildren.slice(from, from + 1)[0];
+        rootChildren.splice(rootChildren.indexOf(item), 1);
+        rootChildren.splice(to, 0, item);
+
+        this.initData(rootChildren);
+        root.setChildren(rootChildren);
+
+        return rootChildren.indexOf(item);
+    }
+
+    getDataByRow(row: number): DATA {
+        const nodes: TreeNode<DATA>[] = this.getRoot().getCurrentRoot().treeToList();
+        const node = nodes[row];
```

Comment:
Codacy found an issue: [Variable Assigned to Object Injection Sink](https://app.codacy.com/manual/enonic/lib-admin-ui/pullRequest?prid=6422415)